### ISSUE TITLE
Console modes do not return changed when using FileDescriptor.noecho! FileDescriptor.echo! FileDescriptor.raw! and FileDescriptor.cooked!

### ISSUE THREAD

**Issue by bmmcginty** [crystal-lang/crystal#14528]
Version: 1.12.1 (2024-04-12) and master

TLDR: cooked! noecho! echo! raw! have their statuses reset before returning to the caller, rendering them no-ops.

```
STDIN.noecho do
puts "password (will not be echoed):"
STDIN.gets
end
STDIN.noecho!
puts "password (will be echoed but should not be):"
STDIN.gets
      ```

The methods in io/console.cr that should permanently alter the console behavior, mainly cooked!, echo!, raw!, and noecho!, fail to work.
I tracked this down to crystal/system/unix/file_descriptor.cr and the system_console_mode function. The system_console_mode function resets the desired modes in an ensure block. The above ! methods in io/console.cr return from the block, but this doesn't stop the code inside the ensure clause from running.

The windows console has similar code, but I don't have the ability to test it.
